File: base64_encoder.py
# base64_encoder.py
import sys
import os
import argparse
import base64
import time
from typing import Optional

def progress_bar(current, total, width=40):
    if total == 0:
        return
    percent = current / total
    filled = int(width * percent)
    bar = '█' * filled + '░' * (width - filled)
    sys.stderr.write(f'\rProgress: [{bar}] {percent*100:.1f}%')
    if current >= total:
        sys.stderr.write('\n')

def encode_file(input_path: str, output_path: Optional[str], url_safe: bool = False):
    if not os.path.exists(input_path):
        print(f"Error: input file '{input_path}' not found.", file=sys.stderr)
        sys.exit(1)
    file_size = os.path.getsize(input_path)
    chunk_size = 1024 * 1024  # 1 MB
    out = open(output_path, 'w') if output_path else sys.stdout
    encoder = base64.urlsafe_b64encode if url_safe else base64.b64encode
    processed = 0
    with open(input_path, 'rb') as f:
        while True:
            chunk = f.read(chunk_size)
            if not chunk:
                break
            encoded = encoder(chunk).decode('ascii')
            out.write(encoded)
            processed += len(chunk)
            progress_bar(processed, file_size)
    if output_path:
        out.close()
    print(f"\n✅ Encoded '{input_path}' to '{output_path or 'stdout'}'")

def decode_file(input_path: str, output_path: Optional[str], url_safe: bool = False):
    if not os.path.exists(input_path):
        print(f"Error: input file '{input_path}' not found.", file=sys.stderr)
        sys.exit(1)
    with open(input_path, 'r') as f:
        data = f.read()
    # Remove whitespace and newlines
    data = ''.join(data.split())
    # Add padding if needed
    missing_padding = len(data) % 4
    if missing_padding:
        data += '=' * (4 - missing_padding)
    try:
        decoder = base64.urlsafe_b64decode if url_safe else base64.b64decode
        decoded = decoder(data)
    except Exception as e:
        print(f"Error: invalid Base64 input - {e}", file=sys.stderr)
        sys.exit(1)
    out = open(output_path, 'wb') if output_path else sys.stdout.buffer
    out.write(decoded)
    if output_path:
        out.close()
    print(f"✅ Decoded '{input_path}' to '{output_path or 'stdout'}'")

def main():
    parser = argparse.ArgumentParser(description="Base64 Encoder/Decoder")
    group = parser.add_mutually_exclusive_group(required=True)
    group.add_argument('-e', '--encode', help='Encode file')
    group.add_argument('-d', '--decode', help='Decode file')
    parser.add_argument('-o', '--output', help='Output file (default: stdout)')
    parser.add_argument('-u', '--url-safe', action='store_true', help='Use URL-safe alphabet')
    args = parser.parse_args()
    if args.encode:
        encode_file(args.encode, args.output, args.url_safe)
    elif args.decode:
        decode_file(args.decode, args.output, args.url_safe)

if __name__ == '__main__':
    main()

File: base64_encoder.go
// base64_encoder.go
package main

import (
	"encoding/base64"
	"flag"
	"fmt"
	"io"
	"os"
	"path/filepath"
)

func progressBar(current, total int64, width int) {
	if total == 0 {
		return
	}
	percent := float64(current) / float64(total)
	filled := int(percent * float64(width))
	bar := ""
	for i := 0; i < filled; i++ {
		bar += "█"
	}
	for i := filled; i < width; i++ {
		bar += "░"
	}
	fmt.Fprintf(os.Stderr, "\rProgress: [%s] %.1f%%", bar, percent*100)
	if current >= total {
		fmt.Fprintln(os.Stderr)
	}
}

func encodeFile(inputPath, outputPath string, urlSafe bool) error {
	info, err := os.Stat(inputPath)
	if err != nil {
		return err
	}
	fileSize := info.Size()
	chunkSize := int64(1024 * 1024) // 1 MB
	out := os.Stdout
	if outputPath != "" {
		out, err = os.Create(outputPath)
		if err != nil {
			return err
		}
		defer out.Close()
	}
	f, err := os.Open(inputPath)
	if err != nil {
		return err
	}
	defer f.Close()
	encoder := base64.StdEncoding
	if urlSafe {
		encoder = base64.URLEncoding
	}
	buf := make([]byte, chunkSize)
	var processed int64
	for {
		n, err := f.Read(buf)
		if n > 0 {
			encoded := encoder.EncodeToString(buf[:n])
			_, werr := out.Write([]byte(encoded))
			if werr != nil {
				return werr
			}
			processed += int64(n)
			progressBar(processed, fileSize, 40)
		}
		if err == io.EOF {
			break
		}
		if err != nil {
			return err
		}
	}
	fmt.Printf("\n✅ Encoded '%s' to '%s'\n", inputPath, outputPath)
	return nil
}

func decodeFile(inputPath, outputPath string, urlSafe bool) error {
	data, err := os.ReadFile(inputPath)
	if err != nil {
		return err
	}
	// Remove whitespace
	clean := make([]byte, 0, len(data))
	for _, b := range data {
		if b != ' ' && b != '\n' && b != '\r' && b != '\t' {
			clean = append(clean, b)
		}
	}
	decoder := base64.StdEncoding
	if urlSafe {
		decoder = base64.URLEncoding
	}
	decoded, err := decoder.DecodeString(string(clean))
	if err != nil {
		return fmt.Errorf("invalid Base64: %w", err)
	}
	out := os.Stdout
	if outputPath != "" {
		out, err = os.Create(outputPath)
		if err != nil {
			return err
		}
		defer out.Close()
	}
	_, err = out.Write(decoded)
	if err != nil {
		return err
	}
	fmt.Printf("\n✅ Decoded '%s' to '%s'\n", inputPath, outputPath)
	return nil
}

func main() {
	var (
		encode   = flag.String("e", "", "Encode file")
		decode   = flag.String("d", "", "Decode file")
		output   = flag.String("o", "", "Output file")
		urlSafe  = flag.Bool("u", false, "Use URL-safe alphabet")
	)
	flag.Parse()
	if *encode == "" && *decode == "" {
		fmt.Fprintln(os.Stderr, "Error: either -e or -d required")
		flag.Usage()
		os.Exit(1)
	}
	var err error
	if *encode != "" {
		err = encodeFile(*encode, *output, *urlSafe)
	} else {
		err = decodeFile(*decode, *output, *urlSafe)
	}
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}
}
